torch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,155 @@
+    return cholesky.view(bmat.shape)
+
+
+def _batch_diag(bmat):
+    """
+    Returns the diagonals of a batch of square matrices.
+    """
+    return bmat.contiguous().view(bmat.shape[:-2] + (-1,))[..., ::bmat.size(-1) + 1]
+
+
+def _batch_mahalanobis(L, x):
+    r"""
+    Computes the squared Mahalanobis distance :math:`\mathbf{x}^\top\mathbf{M}^{-1}\mathbf{x}`
+    for a factored :math:`\mathbf{M} = \mathbf{L}\mathbf{L}^\top`.
+
+    Accepts batches for both L and x.
+    """
+    # TODO: use `torch.potrs` or similar once a backwards pass is implemented.
+    flat_L = L.unsqueeze(0).contiguous().view((-1,) + L.shape[-2:])
+    L_inv = torch.stack([torch.inverse(Li.t()) for Li in flat_L]).view(L.shape)
+    return (x.unsqueeze(-1) * L_inv).sum(-2).pow(2.0).sum(-1)
+
+
+class MultivariateNormal(Distribution):
+    r"""
+    Creates a multivariate normal (also called Gaussian) distribution
+    parameterized by a mean vector and a covariance matrix.
+
+    The multivariate normal distribution can be parameterized either
+    in terms of a positive definite covariance matrix :math:`\mathbf{\Sigma}`
+    or a lower-triangular matrix :math:`\mathbf{L}` with positive-valued
+    diagonal entries, such that
+    :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top`. This triangular matrix
+    can be obtained via e.g. Cholesky decomposition of the covariance.
+
+    Example:
+
+        >>> m = MultivariateNormal(torch.zeros(2), torch.eye(2))
+        >>> m.sample()  # normally distributed with mean=`[0,0]` and covariance_matrix=`I`
+        -0.2102
+        -0.5429
+        [torch.FloatTensor of size 2]
+
+    Args:
+        loc (Tensor or Variable): mean of the distribution
+        covariance_matrix (Tensor or Variable): positive-definite covariance matrix
+        scale_tril (Tensor or Variable): lower-triangular factor of covariance, with positive-valued diagonal
+
+    Note:
+        Only one of `covariance_matrix` or `scale_tril` can be specified.
+
+        Using `scale_tril` will be more efficient: all computations internally
+        are based on `scale_tril`. If `covariance_matrix` is passed instead,
+        this is only used to compute the corresponding lower triangular
+        matrices.
+    """
+    params = {'loc': constraints.real_vector,
+              'covariance_matrix': constraints.positive_definite,
+              'scale_tril': constraints.lower_cholesky}
+    support = constraints.real
+    has_rsample = True
+
+    def __init__(self, loc, covariance_matrix=None, scale_tril=None):
+        event_shape = torch.Size(loc.shape[-1:])
+        if (covariance_matrix is None) == (scale_tril is None):
+            raise ValueError("Exactly one of covariance_matrix or scale_tril may be specified (but not both).")
+        if scale_tril is None:
+            if covariance_matrix.dim() < 2:
+                raise ValueError("covariance_matrix must be two-dimensional")
```

Comment:
Is it always two-dimensional or a batch of two-dimensional matrices?

Reply:
Do you have a suggestion how we should word this? Tentatively updating to:

"scale_tril matrix must be two-dimensional, with optional leading batch dimensions"